athenahealth/httpclient.go

```diff
@@ -386,6 +386,32 @@ func (h *HTTPClient) PostForm(ctx context.Context, path string, v url.Values, ou
 	return h.request(ctx, http.MethodPost, path, body, headers, out)
 }
 
+func (h *HTTPClient) PostFormReader(ctx context.Context, path string, fue *formURLEncoder, out interface{}) (*http.Response, error) {
+	var body io.Reader
+	var headers = http.Header{}
+
+	if fue != nil {
+		pr, pw := io.Pipe()
+
+		go func() {
+			err := fue.Encode(pw)
+			if err != nil {
+				h.logger.Error().Err(err).Msg("encoding url form")
+			}
+
+			err = pw.Close()
+			if err != nil {
+				h.logger.Error().Err(err).Msg("closing pipe writer")
+			}
+		}()
+
+		body = pr
```

Comment:
this is nice, so it can be done asynchronously like that using a pipe?

Reply:
yup this method of piping is pretty powerful and is really the core of what enables the buffering
the pipe reader waits on receiving contents on the pipe writer by using channels and will read all the way up until the pipe writer is closed